### PR TITLE
Improve Paraglide Compiler output

### PR DESCRIPTION
**Pull request by LorisSigrist**
This PR does a few things to improve the Paraglide Compiler & make it more robust.

## Invalid JS identifiers as params
Before, if you used an invalid JS identifier as a variable name the compiler would produce invalid output. Eg:
```js
/**
 * @param {{ 000: NonNullable<unknown> }} params !!!
 * @returns {string}
 */
export const with_nonjs_variable = (params) => `Hallo ${params.000}` //!!!
```

Now any message-variables that aren't valid JS identifiers will be compiled to object-indexes:

```js
/**
 * @param {{ '000' : NonNullable<unknown> }} params
 * @returns {string}
 */
export const with_nonjs_variable = (params) => `Hallo ${params['000']}`
```

## String Escaping
Previously Paraglide didn't do any string-escaping, so if your message included a backtick the compiler would produce invalid output. This PR properly escapes the strings.

## Fail on invalid Message ID
Paraglide already had a half-hearted check for if the message-ID was a valid JS variable name. This PR replaces the check with a robust version.

## Add `/* @__NO_SIDE_EFFECTS__ */` compiler instruction on message functions
The `/* @__NO_SIDE_EFFECTS__ */` comment instructs bundlers to discard the function if it's result is not used. This will result in smaller bundle sizes.

All of these issues bit me when importing the realworld messages from [obsidian-translations](https://github.com/obsidianmd/obsidian-translations)

The valid-JS-Identifier logic is the same as the lint rule, but it has now been extracted into it's own package.

### CODE REVIEW

**Comment by samuelstroschein** [comment]
@LorisSigrist Good changes. You can delete the `valid-js-identifier` lint rule though. 

#1892 will introduce ids that work for paraglide. Aliases then need to be linted with a `valid-paraglide-alias` lint rule. We will delete the `valid-js-identifier` lint rule in January anyways :)

**Review by samuelstroschein — APPROVED** [review]

File: inlang/source-code/paraglide/paraglide-js/src/compiler/compile.ts [hunk 1]
```diff
@@ -70,7 +70,7 @@ export const compile = (args: {
 *! WARNING: Only import messages from this file if you want to manually
 *! optimize your bundle. Else, import from the \`messages.js\` file. 
 * 
-* Your bundler will (in the future) automatically replace the index function
```
[the remark this answers is not in the document]
Comment (LorisSigrist) [reply inlang/source-code/paraglide/paraglide-js/src/compiler/compile.ts:73]: "in future" feels more natural. 
It's the difference between "in der Zukunft" and "künftig"

[the remark this answers is not in the document]
Comment (samuelstroschein) [reply inlang/source-code/paraglide/paraglide-js/src/compiler/compile.ts:73]: Künftig gibt es im englischen nicht. Ich habs mal durch grammarly gejagt. Ohne das "the" ist es inkorrektes Englisch:

CleanShot 2023-12-22 at 14 21 00@2x

**Comment by LorisSigrist** [comment]
The CI was failing because the marketplace's registry `test` script was configured in a way where it depended on it's `build` script running first. This caused a race condition. fixed now

**Pull request merged**